### ISSUE TITLE
Issue with IBControllerGatewayStart.bat and Start TWS Live (daily).xml

### ISSUE THREAD

**Issue by vmitec** [ib-controller/ib-controller#130]
Hello,

Thanks to everyone contributing to ib-controller and making it open source.

I am using ib-controller to perform a daily restart of the IB Gateway as recommended by IB. 

I would like to report an issue I have since I upgraded to version 3.2.0. I am not sure whether the issue is affecting me only or all users, but I believe it is a generic issue.
I cannot get the task "Start TWS Live (daily).xml" to work correctly with IBControllerGatewayStart.bat. 

At the scheduled time, the task starts as expected but does not shut down the existing IB Gateway session and display a pop up "Another session with the same user name already exists. Would you like to login and disconnect the other session or exit application? Continue Login or Exit Application".

In the previous version of ib-controller, the task would have forced IB Gateway to close every 1432 minutes and restarted it.

In addition, as the task is is scheduled to run every 5 minutes, it now opens a command window which closes immediately. In previous version of ib-controller, the task was running in the background, not opening any window. 

I believe the issue has to do with the change made at the end of IBControllerGatewayStart.bat : Before, it was "call Scripts\IBController.bat ..." and in 3.2.0, it has been replaced by "start ... \Scripts\DisplayBannerAndLaunch.bat"
Maybe the start statement starts a new process and this creates issues with the Windows Task scheduler.
I will try to experiment by replacing the "start" by a "call" in  IBControllerGatewayStart.bat and see how it goes. If anybody had the same issue and managed to fix it, please advise.
Thanks

**Comment by rlktradewright** [comment]
You are mistaken. The scheduled task has never forced an already running task to finish. The only things that cause IBController to exit are:

 

*	TWS’s autologooff time has been reached, and IbAutoClosedown is set to Yes (this doesn’t apply to the Gateway)
*	The time configured in Closedown at has been reached
*	IBController receives a Stop command 
*	The user closes TWS/Gateway manually or kills the process
*	A program error occurs in either IBController or the Gateway

 

So if you want to use the scheduled task to start TWS afresh every day (which is what I do), you have to make sure that you have TWS’s autologoff time configured. I have mine set to autologoff at 0542, and the task restarts it at 05:45. That way I keep TWS happy by giving it a daily restart but I miss nothing of any significance.

 

I don’t believe it’s correct to suggest that IB recommend a daily restart of the Gateway. That’s why there is no autologoff feature in the Gateway. Just keep it running as long as you need.

 

The settings in this task should say ‘don’t start the task if another task is already running’, so if you do use it to start the Gateway, it will just keep running until you stop it. Maybe you have changed this setting, and that is the cause of your problem. If not you’ll need to provide a more information about exactly what you’re doing and exactly what’s happening.

 

The command window that opens is important: it contains details of the running TWS or Gateway, and in particular shows where the IBController log file is. That log file is extremely useful for diagnosing problems . So please don’t try to somehow inhibit this window: you’re wasting your time and will probably just dig yourself into a hole. If the presence of the window really bothers you, then change this line in IBControllerGatewayStart.bat:

 

set MIN=

 

to            

 

                set MIN=/Min

 

Then the window will still be there but it will be minimized to the taskbar. The next release of IBController will have a specific option to govern this behaviour.

 

Richard

**Comment by vmitec** [comment]
Thanks Richard for the detailed answer. I will try to run TWS instead of IB Gateways as you do.

Up to now, I always have ran ib-controller with IB Gateway, never with TWS.
I was using the task "Start TWS Live (daily).xml" to restart IB Gateway daily, and this used to work pretty well. 
I just changed the task to run IBControllerGatewayStart.bat instead of IBControllerStart.bat and left all other settings unchanged.

This task contains 2 settings (in Trigger) which used to make  IB Gateway to restart once a day:
- Repeat task every 5 minutes for a duration of 1432 minutes
- Stop task if it runs longer than 1435 minutes

So the daily restart of IB Gateway did not rely on the auto-close feature of TWS.
Since the update of IBControllerGatewayStart.bat, this no longer works: Looks like the Task scheduler is no longer able to stop the task.

Regarding running IB Gateway continuously, I found this was not very reliable unless I was restarting IB Gateway manually at least once a week. 
I had several issues when leaving IB Gateway running continuously, to cite a few: market data interruptions, disconnections from the server, incorrect market hours returned by getContractDetails etc. 

Since I am using ib-controller to perform a daily restart of IB Gateway, all issues are gone. 

Thanks again for your answer, I will try to use TWS with ib-controller instead of IB Gateway. This should solve my issues

**Comment by rlktradewright** [comment]
I've been doing some tests and I now better understand what is going on. Trying to use the scheduled task in its current form with the Gateway is doomed to failure, and it will only work properly with TWS if the autologoff time is suitably configured. 

This is because when the maximum time for the task is exceeded, the Gateway or TWS is not terminated. Even if you remove the start command from the top level command files and call DisplayBannerAndLaunch.bat directly, the cmd session that's running the show isn't terminated when the maximum time is exceeded. So in fact these maximum time settings appear to be useless when the task is running a batch file that in turn is running a program. You can see in the task history that the maximum time has been exceeded, and task scheduler has terminated the task, but the cmd itself just keeps running. (In fact I can't see how it would have been any better in previous IBController versions, but I'll take your word for it: I'm probably missing something obvious.)

As an interim measure, you'd do best to use TWS instead of the Gateway. 

I'll investigate this further and see if I can find a way to improve things.

One possible option for achieving what you want with the Gateway would be to run two scheduled tasks: the first would run IBController to start the Gateway; the second would run at the time you want to shut down the gateway and would send a STOP command to IBController using telnet.

**Comment by vmitec** [comment]
FYI, I have been doing some tests last night and I believe I found a way to solve the issue: 
I just cut & paste the code of DisplayBannerAndLaunch.bat at the end of  IBControllerGatewayStart.bat, in place of the "start ... DisplayBannerAndLaunch.bat".

**Comment by rlktradewright** [comment]
What you suggest doesn't work for me. If it really does work for you, can you please attach your versions of the .bat files and an export of the scheduled task so I can see why it works.

I've also been doing a lot of investigation, and I now understand precisely what is going on and why I think this can never work with the Gateway (at least on Windows 10. It is just possible that Microsoft have changed the behaviour of the command prompt in Windows 10 in such a way that it might work in earlier Windows versions, but it's almost unthinkable that Microsoft would do this: they're extremely reluctant to change things that might impact on existing workloads).

The simple fact is that if you run a program from a command file (ie a .bat file), you can kill the cmd.exe process that processed the command and the running program will continue to run.

When you run a batch file from Task Scheduler, the cmd.exe process that this initiates IS the task, as far as Task Scheduler is concerned. So if the cmd.exe process persists long enough for the maximum task duration to be exceeded, Task Scheduler will kill it, but this doesn't affect any program(s) that the .bat file may have started.

So when running IBGatewayStart.bat from Task Scheduler, it doesn't matter whether IBController is actually started directly from that .bat file or indirectly via a chain of calls, starts or whatever: the only process that is killed when the task's maximum duration is exceeded is the initial cmd.exe process - the java.exe process just keeps running sweetly.

Now, what baffled me and has taken me ages to sort out in my mind is why manually closing the banner window that runs the DisplayBannerAndLaunch.bat causes IBController to stop (hence the warning displayed in the banner). I assumed that when the Task Scheduler task times out, it would similarly kill this window and that would kill IBController in the same way. But there is a subtle difference.

Let me explain my current understanding (please don't feel you need to spend time on what follows unless you're interested).

First of all, it seems reasonable to think that when you start a command prompt session you're only creating a single process, but in fact this is not the case. To demonstrate this, run the Task Manager and show the process list (Processes tab in Windows 7, Details tab in Windows 10). Now start a command prompt: you'll see that a new cmd.exe process appears in the list, as expected, But also a new conhost.exe process appears. This is the Console Window Host process, and is what actually owns the command window that you see on screen. Any console output from the cmd.exe process is piped through to the conhost.exe process for display.

If you run the procexp tool, you'll see that the conhost.exe process is a child of the cmd.exe process, and that the cmd.exe process is a child of the explorer.exe process: I presume cmd.exe actually creates the conhost.exe process itself. (Process Explorer is a Microsoft tool that is like a souped-up Task Manager and shows all sorts of useful information - you can easily find it online.) 

Now run IBControllerGatewayStart.bat (either your modified version with the code for DisplayBannerAndLaunch.bat included inline, or an older version of IBController such as 2.14 with say Gateway v949) within that command prompt. You'll see in procexp that the java.exe process appears as another child process of the cmd.exe process. A java.exe process sends output to the console, and it will use the same conhost.exe process as the cmd.exe process.

If you were to close the command window via the close button in the title bar, this would actually close the console window and thus the conhost.exe process as well as the cmd.exe process, and the loss of the console causes the java.exe process to also die - hence the Gateway will vanish.

But if you just kill the cmd.exe process itself (by right clicking on the cmd.exe process in Task Manager and selecting End Task) this doesn't close the console window, so the banner window remains and the Gateway just keep running fine.

Now what happens when you're running IBControllerStart.bat as a Scheduled Task is that it creates a cmd.exe process to run the command, and this creates the conhost.exe process and the java.exe process. When the maximum task duration is exceeded, the Task Scheduler does exactly the same thing as described previously and kills the cmd.exe process, because that IS the task (I doubt it's even aware of the conhost.exe and java.exe processes at all). So Gateway is left running.

In other words, there is no setting you can make in Task Scheduler to cause it to kill the Gateway java.exe process, because it's knows nothing about it.

**Comment by vmitec** [comment]
This really works for me, trust me! 
Please find attached the the files I have been using for testing (Modified IBControllerGatewayStart.bat + export of the scheduled task). 
With these 2 files, IB Gateway is stopped every 5 minutes  and is restarted 1 minute after.

Can you give it a try and let me know?

[IBControllerGatewayStart.zip](https://github.com/ib-controller/ib-controller/files/758832/IBControllerGatewayStart.zip)

Regarding understanding the way it works. I am far to be a Task Scheduler expert!
However, I noticed 2 things:
* When I manually close the command window opened by the Task Scheduler, the IB Gateway closes too;
* When I stop the running task in the Task Scheduler interface, the command window and the IB Gateway closes.

This seems to indicate that the child processes opened by IBControllerGatewayStart.bat in the command window also get killed when the task scheduler kills the task.

**Comment by vmitec** [comment]
Interesting to note too is that with the original IBControllerGatewayStart.bat, ending the task from the task scheduler (right click on the task then end) has no effect. With the updated IBControllerGatewayStart.bat provided in the .zip file above, ending the task closes the command window+IB Gateway...

**Comment by rlktradewright** [comment]
Well, none of this makes any difference for me. In fact it's even worse than before, because setting the restart interval at 1 minute as you have done (ie less than the maximum duration of 5 minutes) results in Task Scheduler never recording the maximum duration as being reached (I don't know why), let alone then starting a new task with a new instance of the Gateway.

And contrary to what you say, ending the task manually via the Task Scheduler doesn't cause the command window or the Gateway to close. 

Now, I'm running on Windows 10, and as I said before it's possible, though extremely unlikely, that Microsoft have changed the behaviour between whatever Windows version you're using and Windows 10.

But actually, after writing the previous paragraph, I remembered that we have an old Windows XP laptop that somehow hasn't been sent for recycling yet, so I dusted that off, installed TWS and IBController on it, with your modified version of IBControllerStart.bat, and set up a scheduled task to run it (the XP task scheduler doesn't have all the fancy abilities of the recent versions, but it works fine). So I ran the task, the command window appeared followed by the Gateway, and all was well. Then I ended the task manually, the task stopped, but the command window and Gateway just carried on, exactly the same as on Windows 10.

Here's another thing you can do: just run IBControllerGatewayStart.bat (your version) directly by double-clicking in Windows Explorer. In Task Manager, you'll see the cmd.exe and the java.exe processes appear (together with the conhost.exe process with recent Windows versions, but not with XP). Now use Task Manager to kill the cmd.exe process, and even on XP the command window and Gateway keep running, even though the cmd.exe process is gone.

So at the moment, I'm far from convinced that what you're claiming can possibly be true: I'm certainly not trying to say you're telling untruths, just that there's something fishy here that doesn't tally with the way Windows works.

Here are a couple of suggestions that might help to clarify things one way or another.

1. Get your Gateway running via your scheduled task, then look in Task Manager to get the process id (PID) for the java.exe process. After the task has been running for more than 5 minutes, and supposedly restarted, check the PID again. If it's still the same, then the Gateway hasn't been restarted.

2. Check the IBController log file: if the Gateway is indeed being restarted every 5 minutes, then it will be very obvious in the log. If so, please post the log file so I can look at it and see what miracle is happening on your system.

**Comment by vmitec** [comment]
This is weird. I do not understand why it does not work for you. I tried again this morning and it worked like a charm. Per your request, I captured the IB Gateway log file, snapshots of the task manager.
I also added the .bat file and the task I have been using for this test. Btw. I am running this on Windows 7 SP1 64 bits. 

[IBControllerGatewayStart_20170208.zip](https://github.com/ib-controller/ib-controller/files/760768/IBControllerGatewayStart_20170208.zip)

**Comment by rlktradewright** [comment]
Thanks for the files. You're right - this is weird! 

I'm convinced now that it's working on your system, so I need to work out why it isn't working on Windows 10. Well actually I already know why it doesn't work on Windows 10: it's because the Windows 10 Task Scheduler only kills the cmd.exe process when the task's maximum duration is reached, and it is a fact that (at least on Windows 10 and Windows XP) killing that process doesn't stop the conhost.exe process and the java.exe process.

Perhaps you could do an experiment for me? Try just running the IBControllerGatewayStart.bat manually (ie not in a Scheduled Task), and when it's got going, kill the relevant cmd.exe process from Task Manager (don't stop the process tree, just that one process). I'd like to know whether the Gateway keeps running as it does on Windows 10 and Windows XP (I no longer have any Windows 7 machines that I can try this on myself).

Just to make sure I wasn't fooling myself, I reimported your task scheduler XML file, changed the username, and ran it against your copy of IBControllerGAtewayStart.bat. But it still behaves as it did yesterday: ie the 5 minutes max duration is never detected at all, and the task just keeps running. The 1-minute restart attempts are recorded in the Task Scheduler log as 'Launch request ignored, instance already running'. And if I end the task manually, the Gateway just carried on.

So the suspicion I have is that the Windows 7 Task Scheduler kills the process tree when the maximum duration is reached instead of just the cmd.exe process. This is actually much more sensible, and I wonder why they decided to change it in Windows 10? Actually the Task Scheduler has had a rather chequered history during the development of Windows 10, and for a long time during the development of the Anniversary Edition it was full of bugs in the preview versions, but then they seemed to finally get it sorted in time for that release.

But there are a couple of other interesting points. First, when I re-exported the XML file after setting it up and compared it with yours, there is one interesting difference apart from the username. This is that yours has

<UseUnifiedSchedulingEngine>false</UseUnifiedSchedulingEngine>

whereas mine has 

<UseUnifiedSchedulingEngine>true</UseUnifiedSchedulingEngine>

There may be some differences with the unified scheduling engine that account for some of this behaviour, but there's no option to set it to false in Windows 10. 

The other thing this is that I'll try setting the 'Configure for' option to 'Windows 10', rather than its default setting of 'Windows 7, Windows Server 2008 R2' and see if that makes any difference. I've kicked that off but have to be away from the computer for a while now, so I'll let you know later what happens.

**Comment by rlktradewright** [comment]
Well, changing the 'Configure to' option made no difference at all. Not that I really expected it to.

However, some good news, I tried running it on Windows Server 2012 R2, which is the server version of Windows 8.1, and it worked the same as yours.

So this is definitely a Windows 10 problem. I'll have to submit some feedback to the Windows 10 Insider program and hope that Microsoft take some notice. There are several other issues with the Windows 10 Task Scheduler that I haven't touched on here, but I'll feed them back as well. But before I do that I'll have to try this out on the latest insider version - it might have already been fixed.

Anyway, all this leaves me in a bit of a quandary as to the best way to proceed. I'll probably provide a parameter to the Start scripts that results in DisplayBannerAndLaunch.bat being called rather than started (which would make it equivalent to your version, but keeping a nice separation between the bits users can edit and the bits they should leave well alone). I'll have to beef up the description of running IBController with the Task Scheduler. I may even provide a script I knocked up that sends a STOP command to IBController so people running on Windows 10 can shut it down from a separate scheduled task. 

But now it's midnight here in Scotland so I'm off to bed.

**Comment by vmitec** [comment]
Good news ! For me it works fine on Windows 7 SP1 64 bits and on Windows 2008 R2 SP1.
However, IBControllerGatewayStart.bat would need to be updated such as DisplayBannerAndLaunch.bat is called rather than started. 
I was wondering... Did you try to create the task manually in WIndows 10? Maybe the task import/export is not portable between different versions of Windows and results in some incompatibilities.

**Comment by rlktradewright** [comment]
Regarding calling rather than starting DisplayBannerAndLaunch.bat, I mentioned that at the end of my previous post. I want to be able to keep the 'start' option for situations where IBController is not being launched by Task Scheduler. Hence my suggestion of providing a parameter (eg /Inline = don't run in separate window) to the xxxStart.bats.

I've been trying it on the latest Insider Preview version of Windows 10, and it doesn't work there either. I've no idea what Microsoft thinks it's up to with the Windows 10 Task Scheduler. There's masses of feedback about its many faults, but they just don't seem to be addressing the issues.

Creating the task from scratch on Windows 10 makes no difference.

**Comment by rlktradewright** [comment]
A couple of days ago I mentioned that I ran it on XP and it behaved the same way as on Windows 10.

This was not correct: in view of the fact that Windows 7, 8, 8.1 and Server 2012 R2 all behave consistently, I went back to XP and tried it again today, and in fact it doesn't behave like Windows 10, but is the same as all the others.

Not sure what I did that caused that mistake, but it's good to know that Windows 10 is the villain here and has departed from two decades of consistent operation (I know it worked the same in Windows 2000 as well).

**Comment by rlktradewright** [comment]
It wasn't my intention to close this issue the other day, as I don't think it's anything like resolved. I probably just accidentally clicked the 'Close issue' button instead of the 'Comment' button, which is all too easy to do.

I'll only close it when I've made the necessary changes to fix the problem, and I haven't yet come to a decision about the best way to do that. 

Regarding Windows 10, there is currently no way to fix the problem. I might raise a new Windows 10-specific issue, and I'll add something to the User Guide about it.

**Comment by vmitec** [comment]
No problem. I don't know neither what would be the best way to solve this issue. Maybe, IB Controller should not rely on the Windows Task Scheduler nor on IB  TWS auto-restart to perform a daily restart? Could the start-up scripts do this instead?

**Comment by rlktradewright** [comment]
@vmitec thank you for that comment. It reminded me of something I'd been thinking about years ago, but had drifted completely out of my mind, ie running IBController under a Windows service or a Linux daemon.

At the time, the [Apache Commons Daemon](http://commons.apache.org/proper/commons-daemon/index.html) looked like an interesting possibility, and I notice now that there are other options as well. However back then I didn't give it a lot of attention because the Windows Task Scheduler seemed more or less adequate (and cron somewhat less so, but acceptable).

However the situation is now different, with the fiasco of Windows 10's Task Scheduler, and I think it's worth taking another look. 

As it happens there is already some C# code [here](https://github.com/ib-controller/ib-controller/tree/master/resources/IBControllerService) in this repository for running IBController under a service on Windows, contributed by @ShaneCastle a good few years ago. I've never actually done anything with it , partly because I haven't seen the need till now, and partly because it isn't cross-platform. It will be worthwhile having another look at it though.

**Comment by rlktradewright** [comment]
Just to give some sort of closure to this issue, what I've recommended in the User Manual for Windows 10 users wanting to run IBController under Task Scheduler, is to run it first manually and then take the Java command used to start IBController that is recorded in the IBController log file and use that as the Scheduled Task's action. In this way, Java is run directly by Task Scheduler with no batch files involved. 

The downside to this approach is that there is currently no IBController log file when it's run directly in this way. I may do an enhancement for the next release that allows a log file to be passed as a parameter to the IBController program, which would improve things.

I will also give further consideration to running IBController as a service (or rather under the control of a service). A major issue there would be engineering things so that IBController itself would run within an accessible user session.

So I'll now close this issue.
